src/main/java/ca/corefacility/bioinformatics/irida/ria/web/analysis/auditing/AnalysisAuditing.java

```diff
@@ -0,0 +1,117 @@
+package ca.corefacility.bioinformatics.irida.ria.web.analysis.auditing;
+
+/*
+ * This class is used for auditing analysis submissions using
+ * AuditReader and EntityManagerFactory.
+ */
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+
+import org.hibernate.envers.AuditReader;
+import org.hibernate.envers.AuditReaderFactory;
+import org.hibernate.envers.query.AuditEntity;
+
+import ca.corefacility.bioinformatics.irida.model.enums.AnalysisState;
+import ca.corefacility.bioinformatics.irida.model.workflow.submission.AnalysisSubmission;
+import ca.corefacility.bioinformatics.irida.ria.web.utilities.DateUtilities;
+
+public class AnalysisAuditing {
+	private EntityManagerFactory entityManagerFactory;
+
+	public AnalysisAuditing(EntityManagerFactory entityManagerFactory) {
+		this.entityManagerFactory=entityManagerFactory;
+	}
+
+	/**
+	 * Gets the running time of an analysis
+	 *
+	 * @param submission {@link AnalysisSubmission} The submission {@link AnalysisSubmission}
+	 * @return {@link Long} Running time of the analysis
+	 */
+	public Long getAnalysisRunningTime(AnalysisSubmission submission) {
+		EntityManager entityManager = entityManagerFactory.createEntityManager();
+		AuditReader audit = AuditReaderFactory.get(entityManager);
+		ArrayList<AnalysisSubmission> uniqueAuditedSubmissions = new ArrayList<>();
+		if (audit != null ) {
+			// Gets a list of the analysis submission revisions for the submission
+			List<?> auditResultSet = audit.createQuery()
+					.forRevisionsOfEntity(AnalysisSubmission.class, true, false)
+					.add(AuditEntity.id()
+							.eq(submission.getId()))
+					.getResultList();
+			// release the db connection
+			entityManager.close();
+			ArrayList<String> auditedStates = new ArrayList<>();
+
+			// Get a unique list of the audited submissions based on the state
+			for (Object auditResult : auditResultSet) {
+				AnalysisSubmission auditedSubmission = (AnalysisSubmission) auditResult;
+				if (auditedSubmission != null && !auditedStates.contains(auditedSubmission.getAnalysisState()
+						.toString())) {
+					auditedStates.add(auditedSubmission.getAnalysisState()
+							.toString());
+					uniqueAuditedSubmissions.add(auditedSubmission);
+				}
+			}
+			// Get the run time of the analysis from creation till completion/error
+			return DateUtilities.getDurationInMilliseconds(submission.getCreatedDate(),
+					uniqueAuditedSubmissions.get(uniqueAuditedSubmissions.size() - 1)
+							.getModifiedDate());
+		}
+
+		// Before returning we double check to make sure
+		// the connection to the db has been released
+		if(entityManager.isOpen()) {
+			entityManager.close();
```

Comment:
Yeah definitely try to use the AnalysisSubmissionRepository/RevisionRepository.  It'll avoid so much of this stuff.